File: USACO/2.2/castle.cpp
#include <bits/stdc++.h>
using namespace std;
#define loop(i, a) for (int i = 0; i < a; i++)
int M, N;
int castle[51][51];
bool visited[51][51];

int room_count = 0;
int max_room = 0;
int max_removal = 0;
int ans_x = 0;
int ans_y = 0;
char max_wall = 'X';

/*
ID: 11074501
TASK: castle
LANG: C++
*/

int ff(int i, int j, bool removal){
	queue<pair<int,int>> bfs;
	bfs.push(make_pair(i, j));

	bool trav[51][51];
	memset(trav, false, sizeof trav);
	int room_size = 0;

	while(!bfs.empty()){
		pair<int,int> coord = bfs.front();
		bfs.pop();
		int x = coord.first;
		int y = coord.second;
		
		if (trav[y][x] == true){
			continue;
		}

		room_size += 1;

		// no wall to south
		if ( (castle[y][x] & 8) == 0){
			bfs.push(make_pair(x, y+1));
		}

		// no wall to east
		if ( (castle[y][x] & 4) == 0){
			bfs.push(make_pair(x+1, y));
		}

		// no wall to north
		if ( (castle[y][x] & 2) == 0){
			bfs.push(make_pair(x, y-1));
		}

		// no wall to west
		if ( (castle[y][x] & 1) == 0){
			bfs.push(make_pair(x-1, y));
		}

		trav[y][x] = true;
		if (removal == false) {
			visited[y][x] = true;
		}
	}

	return room_size;
}

void consider(int x, int y){

	if (visited[y][x] == false){
		room_count += 1;
		int room_size = ff(x, y, false);
		max_room = max(room_size, max_room);
	}

	bool update = false;
	int test = 0;

	if ( (castle[y][x] & 4) != 0 && x + 1 < M){
		castle[y][x] -= 4;
		test = ff(x, y, true);
		if ((test > max_removal) || 
			(test == max_removal && x < ans_x) || 
			(test == max_removal && x == ans_x && y > ans_y)){
				max_removal = test;
				max_wall = 'E';
				ans_x = x;
				ans_y = y;
			}
		castle[y][x] += 4;
	}

	if ( (castle[y][x] & 2) != 0 && y - 1 >= 0){
		castle[y][x] -= 2;
		test = ff(x, y, true);
		if ((test > max_removal) || 
			(test == max_removal && x < ans_x) || 
			(test == max_removal && x == ans_x && y > ans_y) ||
			(test == max_removal && x == ans_x && y == ans_y && max_wall == 'E')
			){
				max_removal = test;
				max_wall = 'N';
				ans_x = x;
				ans_y = y;
			}
		castle[y][x] += 2;
	}

}


int main(){
	ofstream fout ("castle.out");
	ifstream fin ("castle.in");

	fin >> M >> N;
	memset(castle, 0, sizeof castle);
	memset(visited, false, sizeof visited);
	for (int i = 0; i < N; i++)
		for (int j = 0; j < M; j++)
			fin >> castle[i][j];

	for (int y = 0; y < N; y++){
		for (int x = M-1; x >= 0; x--){
			consider(x, y);
		}
	}

	fout << room_count << endl;
	fout << max_room << endl;
	fout << max_removal << endl;
	fout << ans_y + 1 << " " << ans_x + 1 << " " << max_wall << endl;
}

File: USACO/2.2/castle.py
'''
ID: 11074501
TASK: castle
LANG: PYTHON3
'''
import copy


def ff(x, y, castle, visited, removal):
	bfs = []
	bfs.append((x, y))

	trav = []
	count = 0
	while(len(bfs) > 0):
		i, j = bfs.pop()
		if (i, j) in trav:
			continue

		trav.append((i, j))
		count += 1
		counter = castle[j][i]

		# wall to south
		if (counter & (1 << 3) == 0):
			bfs.append((i, j+1))

		# wall to east
		if (counter & (1 << 2) == 0):
			bfs.append((i+1, j))

		# wall to north
		if (counter & (1 << 1) == 0):
			bfs.append((i, j-1))

		# wall to west
		if (counter & (1 << 0) == 0):
			bfs.append((i-1, j))

	if removal == False:
		for elem in trav:
			i, j = elem
			visited[elem[1]][elem[0]] = count

	return count


def consider(i, j, castle, visited, max_count, max_wall):
	M = len(castle[0])
	N = len(castle)

	increment = 0

	if visited[j][i] == -1:
		# normal ff:
		room_size = ff(i, j, castle, visited, False)
		increment = 1
	else:
		room_size = visited[j][i]

	# remove wall:
	counter = castle[j][i]
	test = 0
	test_x = 0
	test_y = 0

	# wall to east
	if (counter & (1 << 2) != 0 and i + 1 < M):
		castle[j][i] -= 4
		test = ff(i, j, castle, visited, True)
		if test >= max_count:
			max_count = test
			max_wall = 'E'
			test_x = i
			test_y = j
		castle[j][i] += 4

	# wall to north
	if (counter & (1 << 1) != 0 and j - 1 > 0):
		castle[j][i] -= 2
		test = ff(i, j, castle, visited, True)
		if test >= max_count:
			max_count = test
			max_wall = 'N'
			test_x = i
			test_y = j
		castle[j][i] += 2

	'''
	# wall to west
	if (counter & (1 << 0) != 0 and i - 1 > 0):
		castle[j][i] -= 1
		test = ff(i, j, castle, visited, True)
		if test >= max_count:
			max_count = test
			max_wall = 'W'
			test_x = i
			test_y = j
		castle[j][i] += 1

	# wall to south
	if (counter & (1 << 3) != 0 and j + 1 < N):
		castle[j][i] -= 8
		test = ff(i, j, castle, visited, True)
		if test >= max_count:
			max_count = test
			max_wall = 'S'
			test_x = i
			test_y = j
		castle[j][i] += 8
	'''

	return increment, room_size, test, max_wall, test_x, test_y
##############
#a b c # d e #
#f g h # i j #
# ###  #     #
#k#l#m # n o #
### ######   #
#p q r s # t #
#u v w x # z #
##############

with open('castle.in', 'r') as f:
	thing = f.readline().split()
	M = int(thing[0])
	N = int(thing[1])

	castle = []
	for i in range(N):
		thing = f.readline().split()
		castle.append([int(i) for i in thing])

	visited = [[-1]*M for i in range(N)]

	num_rooms = 0
	max_room_size = 0
	max_count = 0
	max_wall = ''
	ans_x = 0
	ans_y = 0

	for x in range(M-1, -1, -1):
		for y in range(N):
			inc, room_size, test_count, test_wall, test_x, test_y = consider(x, y, castle, visited, max_count, max_wall)
			num_rooms += inc
			max_room_size = max(max_room_size, room_size)
			if test_count > max_count or (test_count == max_count and test_x < ans_x) or (test_count == max_count and test_x == ans_x and test_y > ans_y):
				max_count = test_count
				max_wall = test_wall
				ans_x = test_x
				ans_y = test_y

	with open('castle.out', 'w') as fo:
		fo.write(str(num_rooms))
		fo.write('\n')
		fo.write(str(max_room_size))
		fo.write('\n')
		fo.write(str(max_count))
		fo.write('\n')
		fo.write(str(ans_y + 1) + ' ' + str(ans_x + 1) + ' ' + str(max_wall))
		fo.write('\n')